content/en/blog/_posts/2023-12-13-scheduler-queueinghint/index.md

```diff
@@ -0,0 +1,111 @@
+Here QueueingHints come into play.
+
+## QueueingHint
+
+QueueingHint is callback function per plugin to notice an object addition/update/deletion in the cluster (we call them cluster events)
+that may make Pods schedulable.
+
+Let's say the Pod `pod-a` has a required Pod affinity, and got rejected in scheduling cycle by the `PodAffinity` plugin
+because no Node has any Pod matching the Pod affinity specification for `pod-a`.
+
+![pod-a got rejected by PodAffinity](./queueinghint1.svg)
+
+When an unscheduled Pod is put into the unschedulable pod pool, the scheduling queue
+records which plugins caused the scheduling failure of the Pod.
+In this example, scheduling queue notes that `pod-a` was rejected by PodAffinity.
+
+`pod-a` will never be schedulable until the PodAffinity failure is resolved somehow.
+The Scheduling Queue uses the QueueingHints from plugins that rejected the Pod, which is PodAffinity in the example.
+
+A QueueingHint subscribes to a particular kind of cluster event and make a decision whether an incoming event could make the Pod schedulable.
+Thinking about when PodAffinity failure could be resolved,
+one possible scenario is that an existing Pod gets a new label which matches with `pod-a`'s PodAffinity.
+
+PodAffinity QueueingHint checks all Pod updates happening in the cluster, 
+and when it catches such update, the scheduling queue moves `pod-a` to activeQ/backoffQ.
+
+![pod-a is moved by PodAffinity QueueingHint](./queueinghint2.svg)
+
+We actually already have a similar functionality called `preCheck` inside the scheduling queue,
+which filter out cluster events based on Kubernetes core scheduling constraints - 
+for example, filtering out node related events when nodes aren't ready.
+
+But, it's not ideal because this hard-coded `preCheck` refers to in-tree plugins logic,
+and it causes issues for custom plugins (for example: [#110175](https://github.com/kubernetes/kubernetes/issues/110175)).
+
+## What's new in v1.29
+
+Within SIG Scheduling, we have been working on the development of QueueingHint since
+Kubernetes v1.28.
+In v1.28, only one alpha plugin (DRA) supported QueueingHint,
+and in v1.29, some stable plugins started to implement QueueingHints.
+
+QueueingHint is not something user-facing, but we have a feature gate (`SchedulerQueueingHints`) as a safety net 
+because QueueingHint changes a critical path of the scheduler and adds some memory overhead, depending on how busy a cluster is.
+
+## Getting involved
+
+These features are managed by Kubernetes [SIG Scheduling](https://github.com/kubernetes/community/tree/master/sig-scheduling).
+
+Please join us and share your feedback. 
+
+## How can I learn more?
+
+- [KEP-4247: Per-plugin callback functions for efficient requeueing in the scheduling queue](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/4247-queueinghint/README.md)
```

Comment:
I hope that for beta, or maybe GA, there is a document about how to use this in your custom scheduler that extends kube-scheduler.

Reply:
IMO, that's beyond the scope of user docs.

Reply:
Hmm.

If this isn't something we'll document: why is it important enough to announce in a blog article?

Reply:
I also think how to use is too much, while an overview is worthwhile for everyone.

Not only QueueingHint, but the scheduling framework in general is on the border between internal and user-facing. It's a very fundamental design and it could be important not only for custom-scheduler developers but also for everyone to understand how K8s scheduling works. So, we have a doc explaining an overview of the scheduling framework and each interface: https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/
But, OTOH, we don't explain how to actually implement each interface in detail (e.g., Filter can return these statuses, etc), while the actual interfaces have comments explaining such detail (so, does QueueingHint). This is how the scheduling framework is explained currently, and IMO it's a good balance - If we explain everything about each interface in K8s doc, it'd be too huge to maintain and could be rotten very easily. Instead, comments in source code could be easily maintained with the implementation.

Talking about this blog post, it's also an overview of what QueueingHint does (it's like [this doc](https://github.com/kubernetes/website/pull/43813) explains what each interface does, but a bit more detailed overview), but we deliberately don't go into an implementation detail or how the users can implement their QueueingHint for their custom schedulers.